mozci/console/commands/push.py

```diff
@@ -314,6 +339,76 @@ def _serialize_regressions(regressions):
                     emails, matrix_room, push, previous, classification, regressions
                 )
 
+    def retrigger_failures(self, groups, count, allowed_patterns, retrigger_limit):
+        groups_with_failures = {}
+        for name, failing_tasks in groups.items():
+            filtered_failing_tasks = [
+                task
+                for task in failing_tasks
+                if any(
+                    fnmatch.fnmatch(task.label, pattern) for pattern in allowed_patterns
+                )
+            ]
+            if filtered_failing_tasks:
+                assert all(
+                    any(
+                        not result.ok and result.group == name
+                        for result in task.results
+                    )
+                    for task in filtered_failing_tasks
+                ), f"Some failing tasks on the group {name} (to be retriggered) didn't really fail"
+                groups_with_failures[name] = filtered_failing_tasks
+
+        if not groups_with_failures:
+            return
+
+        for failing_tasks in itertools.islice(
+            groups_with_failures.values(), 0, retrigger_limit
+        ):
+            # If there is more than one task failing in this group, we should retrigger only one of them
+            for _ in range(0, count):
+                failing_tasks[0].retrigger()
+
+    def backfill_and_retrigger_failures(
+        self,
+        push: Push,
+        allowed_patterns: List[str],
+        classify_parameters: Dict[str, Any],
+        retrigger_limit: int,
+        backfill_limit: int,
+        to_retrigger_or_backfill: ToRetriggerOrBackfill,
+    ) -> None:
+        # Retrigger real failures
+        self.retrigger_failures(
+            to_retrigger_or_backfill.real_retrigger,
+            classify_parameters.get("consistent_failures_counts", (2, 3))[1],
+            allowed_patterns,
+            retrigger_limit,
+        )
+
+        # Retrigger intermittent failures
+        self.retrigger_failures(
+            to_retrigger_or_backfill.intermittent_retrigger,
+            classify_parameters.get("consistent_failures_counts", (2, 3))[0],
+            allowed_patterns,
+            retrigger_limit,
+        )
+
+        # Backfill some failures
+        groups_to_backfill = {
+            name: failing_tasks
+            for name, failing_tasks in to_retrigger_or_backfill.backfill.items()
+            if failing_tasks
+        }
+        for failing_tasks in itertools.islice(
+            groups_to_backfill.values(), 0, backfill_limit
+        ):
+            for t in failing_tasks:
+                if t.label and any(
+                    fnmatch.fnmatch(t.label, pattern) for pattern in allowed_patterns
+                ):
+                    t.backfill(push)
```

Comment:
@Archaeopteryx should we notify the group on Matrix when we trigger a backfill?

Reply:
No, failures in the backfills is what is important.